# Transforming Text into Features Using Word Embeddings

In this demo, you will see how to train a Word2Vec model to obtain word embeddings for spam message classification. We will train a logistic regression model using the embeddings to create our feature vectors. We will then use our model to predict whether a new email is spam.

A Word2Vec model should be trained on a large text corpus that is relevant to the domain in which you will later use a machine learning model to make predictions. For example, if you will be working with medical data, your Word2Vec model should not be trained on text data originating from product reviews, but rather should be trained on medical data.

You can train a Word2Vec model using an existing tool, such as Gensim, which is a Python package used for natural language processing. After the embeddings have been produced, you will use them to create feature vectors for your machine learning model. For example, we can represent each training example as a feature vector by taking the average of the Word2Vec embeddings of the words in that training example. Once we have created feature vectors for all of our training examples, we can fit our machine learning model to the training data and use our trained model to make predictions. 

This demo we will walk you through the following steps:

1. Load the spam message data set. The spam email dataset contains email subject lines.

2. Create labeled examples out of the data: We will have one text feature and one label. Each feature will contain one email subject line.

3. Preprocess the text features: We will preprocess the data by removing stop words, converting all text to lowercase, removing punctuation, etc. from the email subject lines. 

4. Create training and test datasets: Each example in the training and test datasets will now contain one preprocessed email subject line as a feature.

5. Train the Word2Vec model using the training dataset's features: we will train a Word2Vec model using Gensim. The model will figure out how to represent words as vectors based on how they are used in the spam email dataset. You will inspect the resulting word embeddings to develop a better understanding of what they can tell you about a particular word.

6. Create feature vectors out of the training and test data: After training the Word2Vec model, we will represent the features in every training and test example (recall, each example contains an email subject line) as a vector by taking the average of the Word2Vec embeddings of the words in the example. 

7. Train a logistic regression model using the feature vectors as input features, and evaluate the model's performance.

**<font color='red'>Note: Some of the code cells in this notebook may take a while to run.</font>**

### Import Packages

Before you get started, import a few packages. Run the code cell below. 

In [1]:
import pandas as pd
import numpy as np
import os 
import matplotlib as plt
import seaborn as sns

We will also import the scikit-learn `LogisticRegression class`, the `train_test_split()` function for splitting the data into training and test sets, and the `roc_auc_score()` function to evaluate the model.


In [2]:
from sklearn.model_selection import train_test_split
from sklearn.linear_model import LogisticRegression
from sklearn.metrics import roc_auc_score

## Step 1: Build Your DataFrame and Define Your ML Problem

We will work with a dataset containing emails that are labeled as either spam or not spam.

#### Load a Data Set and Save it as a Pandas DataFrame

In [3]:
filename = os.path.join(os.getcwd(), "data", "spamDataset.csv")
df = pd.read_csv(filename, header=0)

df.head()

email_text   spam
0  Subject: enron methanol ; meter # : 988291\r\n...  False
1  Subject: hpl nom for january 9 , 2001\r\n( see...  False
2  Subject: neon retreat\r\nho ho ho , we ' re ar...  False
3  Subject: photoshop , windows , office . cheap ...   True
4  Subject: re : indian springs\r\nthis deal is t...  False

#### Define the Label

This is a binary classification problem in which we will predict whether an email is a spam or not spam. The label is the `spam` column.

#### Identify Features

We only have one feature. The feature is the `email_text` column.


## Step 2: Create Labeled Examples from the Data Set

Let's create labeled examples from our dataset. We will have one text feature and one label. 
The code cell below carries out the following steps:

* Gets the `spam` column from DataFrame `df` and assigns it to the variable `y`. This will be our label. Note that the label contains True or False values that indicate whether an email is spam or not spam.
* Gets the column `email_text` from DataFrame `df` and assigns it to the variable `X`. This will our feature. Note that the `email_text` feature contains the subject line of the email.


In [4]:
y = df['spam'] 
X = df['email_text']

X.shape

(5170,)

In [5]:
X.head()

0    Subject: enron methanol ; meter # : 988291\r\n...
1    Subject: hpl nom for january 9 , 2001\r\n( see...
2    Subject: neon retreat\r\nho ho ho , we ' re ar...
3    Subject: photoshop , windows , office . cheap ...
4    Subject: re : indian springs\r\nthis deal is t...
Name: email_text, dtype: object

Let's take a look at an example of a spam email and an email that is not spam.

In [6]:
print('A spam email: \n\n', X[67])
print('A non-spam email: \n\n', X[135])

A spam email: 

 Subject: re : husband soup would be
as you know election time is not the best thing for the economy .
economy is in a very unstable condition , as you can see gas prices
are going up along with the m o rtgvage rat e s . once the
r a te goes up you will not have a chance to s av e money again
for a very long time .
it is your last chance . get r e f inanced at 4 . 2 point !
http : / / www . fintod . com /
- -
despoil , compote a amende
the me orbital irruption
gfawn a ax henrietta
a the in boatswain
out whither the accompanist lint macintosh

A non-spam email: 

 Subject: re : tuesday , december 26 th
i will be here tuesday , also .
mark mccoy
12 / 20 / 2000 09 : 04 am
to : michael olsen / na / enron @ enron , tom acton / corp / enron @ enron , clem
cernosek / hou / ect @ ect , robert cotten / hou / ect @ ect , jackie young / hou / ect @ ect ,
sabrae zajac / hou / ect @ ect , carlos j rodriguez / hou / ect @ ect , mark
mccoy / corp / enron @ enron , aimee lannou / hou /

## Step 3: Preprocess the Text

The next step is to preprocess the text. Preprocessing technqiues can include cleaning the data, converting all text to lowercase, removing special characters, removing stopwords from the text, tokenizing the text (split it into smaller chunks), and lemmatizing the text (converting a word to its root word).

You can perform preprocessing on your own or use tools to accomplish this. One common tool is NLTK [Natural Language Toolkit](https://www.nltk.org/). However, for this demo, we will use the built-in function from Gensim to preprocess the text. This function will remove some stop words, covert all text to lowercase, remove punctuation and tokenize the text.

Let's import the Gensim package.

In [7]:
import gensim

Before we preprocess our data, let's look at an example output of preprocessed text. Let's take a simple sentence and perform preprocessing. Run the code cell below and inspect the results.

In [8]:
sentence = "I went to the market to buy some apples for my pet horse." 

list(gensim.utils.simple_preprocess(sentence))

['went',
 'to',
 'the',
 'market',
 'to',
 'buy',
 'some',
 'apples',
 'for',
 'my',
 'pet',
 'horse']

Now let's perform preprocessing on our `email_text` feature and compare the difference between the original text and the preprocessed text

In [9]:
original_X = X
X = X.apply(lambda row: gensim.utils.simple_preprocess(row))

In [10]:
X.head()

0    [subject, enron, methanol, meter, this, is, fo...
1    [subject, hpl, nom, for, january, see, attache...
2    [subject, neon, retreat, ho, ho, ho, we, re, a...
3    [subject, photoshop, windows, office, cheap, m...
4    [subject, re, indian, springs, this, deal, is,...
Name: email_text, dtype: object

In [11]:
original_X.head()

0    Subject: enron methanol ; meter # : 988291\r\n...
1    Subject: hpl nom for january 9 , 2001\r\n( see...
2    Subject: neon retreat\r\nho ho ho , we ' re ar...
3    Subject: photoshop , windows , office . cheap ...
4    Subject: re : indian springs\r\nthis deal is t...
Name: email_text, dtype: object

## Step 4: Create Training and Test Data Sets

Let's take our preprocessed text data set and split the data into training and test sets with 80% of the data being the training set.

In [12]:
X_train, X_test, y_train, y_test = train_test_split(X, y, train_size=.80, random_state=1234)

X_train.head()

562     [subject, hpl, nom, for, sept, see, attached, ...
148     [subject, copanno, changes, forwarded, by, ami...
4831    [subject, re, maynard, oil, revised, nom, dare...
3385    [subject, deal, for, december, can, either, of...
2389    [subject, heisse, sx, action, hallo, mein, lie...
Name: email_text, dtype: object

## Step 5: Training the Word2Vec Model and Inspecting the Word Embeddings

Now that the data has been preprocessed and we have our training data, we will train a Word2Vec model using Gensim using the training data `X_train`. For more information on Gensim, consult the online [documentation](https://radimrehurek.com/gensim/models/word2vec.html). The model will produce word embeddings. That is, words represented as numerical vectors based on how these words are used in the email spam dataset.
    
We will specify the following parameters:

* size: dimension of each resulting vector that will represent each word.
* window: the number of words behind or ahead of a target word that will be used to provide context for that word
* min_count: the number of times a word must appear in our text document in order to create a word vector. The model will ignore words that do not satisfy the `min_count` specification, therefore ignoring wors that are not important.



In [13]:
print("Begin")
word2vec_model = gensim.models.Word2Vec(X_train,
                                   vector_size=100,
                                   window=5,
                                   min_count=2)

print("End")

Begin
End


You can check the size of a particular model with the `len()` method. 

In [14]:
len(word2vec_model.wv.key_to_index)  # retrieve vocabulary and measure its size

18015

While 18,015 words may seem like a lot, this is actually quite small for most real-world applications. In practice, you can use models with larger vector sizes (say 300) and much larger vocabulary.

To check if a word is included the vocabulary, you can use the `key_to_index` attribute. 

In [15]:
'cornell' in word2vec_model.wv.key_to_index

False

In [16]:
'dog' in word2vec_model.wv.key_to_index

True

The word `'dog'` is in our vocabulary. Let's inspect its word vector. The code cell belows retrieves the vector for the word `'dog'`. 

In [17]:
word2vec_model.wv['dog']

array([-0.072024  ,  0.21465424,  0.00767706, -0.18751048,  0.03749982,
       -0.16994251,  0.0945733 ,  0.4324272 , -0.26145768, -0.19040789,
       -0.21077152, -0.34527636, -0.02158943, -0.03863806,  0.15579635,
       -0.27335536,  0.1021451 , -0.14785983, -0.01033554, -0.35376737,
        0.14145418,  0.06183926,  0.1839573 , -0.17079215, -0.01018147,
        0.06903811, -0.19979025, -0.12217257, -0.24092546,  0.0029037 ,
        0.11197709, -0.04537754,  0.09216053, -0.27917367, -0.12177479,
        0.08432655, -0.09426836, -0.13510141, -0.15318795, -0.29001006,
        0.0637412 , -0.03756299, -0.17595664,  0.05110542,  0.24191   ,
       -0.14005563,  0.0270287 , -0.20837867,  0.11565524,  0.24776776,
        0.03839585, -0.09443516, -0.25166667, -0.10773394, -0.21685329,
        0.02997843, -0.0531922 , -0.00468927, -0.21950328,  0.03823226,
        0.0683376 ,  0.02594218,  0.04891578,  0.00134473, -0.2496334 ,
        0.16943933,  0.2322774 ,  0.21151468, -0.29574957,  0.22

Examine the vector, noting each of the following important characteristics: 

1. It contains 100 values somewhere between -0.39333 and 0.4495
2. There are no zeros, so it can be considered a **dense vector** (a vector comprised of mostly non-zero values)
3. Each value represents a dimension, but is not necessarily interpretable by humans
4. Larger (in magnitude) values may relate to categories in which `'dog'` has a presence
5. Many broader categories are also likely to be represented by a combination of these dimensions
6. A few values are relatively close to zero. This implies that these dimensions do not carry "significant information" about `'dog'`


Let's find the most similar words for `'dog'`.


In [18]:
word2vec_model.wv.most_similar('dog')

[('filthy', 0.9938415884971619),
 ('mi', 0.9923204183578491),
 ('cholesterol', 0.9919399619102478),
 ('und', 0.9913329482078552),
 ('everywhere', 0.9912003874778748),
 ('backlog', 0.9911063313484192),
 ('deliquescent', 0.9910830855369568),
 ('vous', 0.9907364845275879),
 ('von', 0.9905772805213928),
 ('ado', 0.9905033111572266)]

The code cell below outputs the first few 25 words that our Word2Vec model learned a vector for.

In [19]:
top25 = word2vec_model.wv.index_to_key[:25]
top25

['the',
 'to',
 'ect',
 'and',
 'for',
 'of',
 'you',
 'subject',
 'in',
 'on',
 'is',
 'this',
 'hou',
 'enron',
 'be',
 'that',
 'we',
 'from',
 'your',
 'will',
 'have',
 'with',
 'at',
 'are',
 'it']

You can go a step further and print vectors associated with each of these words. Use the `background_gradient` method to wrap these vectors into a DataFrame as rows with the corresponding words as row indices. Let's spice it up with colors.

In [20]:
pd.DataFrame({w:word2vec_model.wv[w] for w in top25}).T.style.background_gradient(cmap='coolwarm').set_precision(2)

Note that the 100 columns you see here are dimensions you specified earlier. We do not know what each dimension signifies, although this is an active area of research. However, if you examine each dimension you'll notice more red in some columns and more blue in others. Some of the words are very common and might be stop words we would normally remove. However, if we load more specific groups of words, we might make good guesses about what some of the dimensions mean.

The code cells below compute the similarities between two words in the vocabulary.

In [21]:
word2vec_model.wv.similarity(w1='dog', w2='person')

0.37227347

In [22]:
word2vec_model.wv.similarity(w1='dog', w2='cat')

0.969508

## Step 6: Create Feature Vectors out of Word Embeddings for a Classifier

Now let's convert the features in our training and test datasets into feature vectors using our word embeddings. We will use this  to train a logistic regression model. Let's first inspect our original training and test datasets:

In [23]:
X_train.head()

562     [subject, hpl, nom, for, sept, see, attached, ...
148     [subject, copanno, changes, forwarded, by, ami...
4831    [subject, re, maynard, oil, revised, nom, dare...
3385    [subject, deal, for, december, can, either, of...
2389    [subject, heisse, sx, action, hallo, mein, lie...
Name: email_text, dtype: object

In [24]:
X_test.head()

5075    [subject, hpl, nom, for, may, see, attached, f...
3817    [subject, cleburne, tenaska, iv, plant, daren,...
2967    [subject, re, producer, connects, on, the, con...
3014    [subject, mail, hey, daren, is, this, your, ma...
2185    [subject, revision, forest, oil, november, gas...
Name: email_text, dtype: object

In [43]:
print(X_train[562])

['subject', 'hpl', 'nom', 'for', 'sept', 'see', 'attached', 'file', 'hplo', 'xls', 'hplo', 'xls']


In [45]:
print(X_test[5075])

['subject', 'hpl', 'nom', 'for', 'may', 'see', 'attached', 'file', 'hplno', 'xls', 'hplno', 'xls']


In [46]:
print(X_train[148])
print(X_test[3817])

['subject', 'copanno', 'changes', 'forwarded', 'by', 'ami', 'chokshi', 'corp', 'enron', 'on', 'am', 'troy', 'benoit', 'reliantenergy', 'com', 'on', 'am', 'to', 'ami', 'chokshi', 'corp', 'enron', 'enron', 'cc', 'subject', 'copanno', 'changes', 'see', 'attached', 'file', 'hpl', 'jan', 'xls', 'hpl', 'jan', 'xls']
['subject', 'cleburne', 'tenaska', 'iv', 'plant', 'daren', 'trying', 'to', 'put', 'together', 'the', 'operating', 'budget', 'for', 'the', 'plant', 'need', 'conservative', 'forecast', 'per', 'month', 'of', 'the', 'price', 'that', 'could', 'be', 'obtained', 'by', 'selling', 'our', 'excess', 'volumes', 'thanks']


The code cell below walks through every example in both the training and test datasets and replaces every word contained in the `email_text` feature with its corresponding word embedding. Original words that do not have a corresponding word embedding will not be part of the training and test sets. For example, stop words that were removed when creating the word embeddings will not appear in the training and test sets.

<b>Note</b>: This may take a while to run.

In [25]:
words = set(word2vec_model.wv.index_to_key)

print('Begin transforming X_train')
X_train_word_embeddings = np.array([np.array([word2vec_model.wv[word] for word in words if word in training_example])
                        for training_example in X_train], dtype=object)
print('Finish transforming X_train')

print('Begin transforming X_test')
X_test_word_embeddings = np.array([np.array([word2vec_model.wv[word] for word in words if word in training_example])
                        for training_example in X_test], dtype=object)
print('Finish transforming X_test')


Begin transforming X_train
Finish transforming X_train
Begin transforming X_test
Finish transforming X_test


In [26]:
print('Number of words in first training example: {0}'.format(len(X_train.iloc[0])))
print('First word in first training example: {0}'.format(X_train.iloc[0][0]))
print('Second word in first training example: {0}\n'.format(X_train.iloc[0][1]))

print('Number of word vectors in first training example: {0}'.format(len(X_train_word_embeddings[0])))
print('First word vector in first training example:\n {0}'.format(X_train_word_embeddings[0][0]))
print('\nSecond word vector in first training example: \n {0}\n'.format(X_train_word_embeddings[0][1]))


Number of words in first training example: 12
First word in first training example: subject
Second word in first training example: hpl

Number of word vectors in first training example: 10
First word vector in first training example:
 [-1.5989214  -0.11416469 -0.81175286 -0.9274372  -0.1274449  -1.0157328
  0.37283322  1.4337475  -0.17774874 -0.5856301  -0.33479342 -0.9088201
  0.34105337 -0.73547095 -0.41861287 -0.14808597 -0.1133917  -0.65034145
  0.88657916 -0.15759015  2.4625204   1.452959   -0.32489064 -0.24212623
 -0.33692425  0.73604995 -1.9165487  -0.93607277 -0.62739676 -0.27604765
 -1.7383752   1.387022    0.5695261  -0.04800386  1.5895836   1.9005555
 -0.61044407  0.32349145  0.72487867 -1.0496564  -0.36158243  1.3147554
 -0.8170927   0.79048616  1.6462389  -1.4739649  -0.39625445 -0.34474063
  0.3138335   0.602399    0.670839    0.11051651  0.03698491 -0.11396274
  1.1590644   1.2047292   1.4409007   0.10784645 -1.1452628   0.28726387
 -1.227442    1.1280564   0.01048497 -0

After replacing the `email_text` feature in our training and test data with word embeddings, each example in our training and test data now has a different number of features, each corresponding to a word vector:

In [27]:
print('Number of word vectors in first five examples in training set:')
for w in range(0, 5):
    print(len(X_train_word_embeddings[w]))

print('Number of word vectors in first five examples in test set:')
for w in range(0, 5):
    print(len(X_test_word_embeddings[w]))

Number of word vectors in first five examples in training set:
10
23
37
57
30
Number of word vectors in first five examples in test set:
10
31
59
8
59


In [39]:
print(X_train_word_embeddings[0])

[[-1.59892142e+00 -1.14164688e-01 -8.11752856e-01 -9.27437186e-01
  -1.27444893e-01 -1.01573277e+00  3.72833222e-01  1.43374753e+00
  -1.77748740e-01 -5.85630119e-01 -3.34793419e-01 -9.08820093e-01
   3.41053367e-01 -7.35470951e-01 -4.18612868e-01 -1.48085967e-01
  -1.13391705e-01 -6.50341451e-01  8.86579156e-01 -1.57590151e-01
   2.46252036e+00  1.45295894e+00 -3.24890643e-01 -2.42126226e-01
  -3.36924255e-01  7.36049950e-01 -1.91654873e+00 -9.36072767e-01
  -6.27396762e-01 -2.76047647e-01 -1.73837519e+00  1.38702202e+00
   5.69526076e-01 -4.80038598e-02  1.58958364e+00  1.90055549e+00
  -6.10444069e-01  3.23491454e-01  7.24878669e-01 -1.04965639e+00
  -3.61582428e-01  1.31475544e+00 -8.17092717e-01  7.90486157e-01
   1.64623892e+00 -1.47396493e+00 -3.96254450e-01 -3.44740629e-01
   3.13833505e-01  6.02398992e-01  6.70839012e-01  1.10516511e-01
   3.69849093e-02 -1.13962740e-01  1.15906441e+00  1.20472920e+00
   1.44090068e+00  1.07846454e-01 -1.14526284e+00  2.87263870e-01
  -1.22744

In [40]:
print(len(X_train_word_embeddings[0]))

10


This will cause an error when we train our model. We have to create feature vectors that will provide our classifier with a consistent set of features per example.

We can take an element-wise average of the word embeddings of the words contained in each training and test example. This makes feature vector representations that can be used as training and test features for our classifier. 

In [28]:
X_train_feature_vector = []
for w in X_train_word_embeddings:
    if w.size:
        X_train_feature_vector.append(w.mean(axis=0))
    else:
        X_train_feature_vector.append(np.zeros(100, dtype=float))
        
X_test_feature_vector = []
for w in X_test_word_embeddings:
    if w.size:
        X_test_feature_vector.append(w.mean(axis=0))
    else:
        X_test_feature_vector.append(np.zeros(100, dtype=float))

In [33]:
print(X_train_feature_vector)

[array([-1.3224926 ,  0.14386567, -0.7610941 ,  0.01526944, -0.46643615,
        -0.4344024 ,  0.17006335,  0.8957881 ,  0.08238375, -0.7400457 ,
        -0.05810558, -0.6272258 , -0.47121078, -0.7206644 , -0.4232912 ,
         0.33997124,  0.0671952 , -0.01619555,  0.8779928 , -0.8748271 ,
         1.8396393 ,  0.72074425,  0.32201645, -0.30848464, -0.1108043 ,
         0.77230394, -1.0076593 , -0.63022506, -0.5204911 ,  0.05925214,
        -1.0533293 ,  1.45436   ,  0.84057677, -0.54644847,  1.4716123 ,
         1.0961179 , -0.58778393,  0.4919681 ,  0.0245184 , -1.0460684 ,
         0.18138416,  0.12316506, -1.1439185 ,  0.23244223,  1.5143749 ,
        -1.7644119 , -0.07761798, -0.6919662 ,  0.95313805,  0.30203348,
         0.4794467 , -0.14340606,  0.22965412,  0.5047895 ,  0.5163913 ,
         1.2116988 ,  0.921024  ,  0.06492339, -1.0178838 ,  0.08445159,
        -0.2301239 ,  0.5891973 , -0.17013371, -0.3523721 , -1.4264389 ,
        -0.03570082, -0.44175577,  0.6625608 , -0.9

Each example now consists of one feature, which is a numerical feature vector of length 100. Run the code cell below to inspect the first five training examples.

In [29]:
for w in range(0, 5):
    print('Length of training example {0}: {1}'.format(w, len(X_train_feature_vector[w])))
    
print('First training example\'s feature vector: \n{0}'.format(X_train_feature_vector[0]))

Length of training example 0: 100
Length of training example 1: 100
Length of training example 2: 100
Length of training example 3: 100
Length of training example 4: 100
First training example's feature vector: 
[-1.3224926   0.14386567 -0.7610941   0.01526944 -0.46643615 -0.4344024
  0.17006335  0.8957881   0.08238375 -0.7400457  -0.05810558 -0.6272258
 -0.47121078 -0.7206644  -0.4232912   0.33997124  0.0671952  -0.01619555
  0.8779928  -0.8748271   1.8396393   0.72074425  0.32201645 -0.30848464
 -0.1108043   0.77230394 -1.0076593  -0.63022506 -0.5204911   0.05925214
 -1.0533293   1.45436     0.84057677 -0.54644847  1.4716123   1.0961179
 -0.58778393  0.4919681   0.0245184  -1.0460684   0.18138416  0.12316506
 -1.1439185   0.23244223  1.5143749  -1.7644119  -0.07761798 -0.6919662
  0.95313805  0.30203348  0.4794467  -0.14340606  0.22965412  0.5047895
  0.5163913   1.2116988   0.921024    0.06492339 -1.0178838   0.08445159
 -0.2301239   0.5891973  -0.17013371 -0.3523721  -1.4264389  -0

## Step 7: Fit a Logistic Regression Model to the Training Data and Evaluate the Model

Now we can train our model on our transformed data. The code cell below trains a logistic regression model and computes the AUC on the test set.

In [30]:
# 1. Create a LogisticRegression model object, and fit a Logistic Regression model to the transformed training data
model = LogisticRegression(max_iter=200)
model.fit(X_train_feature_vector, y_train)

# 2. Make predictions on the transformed test data using the predict_proba() method and 
# save the values of the second column
probability_predictions = model.predict_proba(X_test_feature_vector)[:,1]

# 3. Make predictions on the transformed test data using the predict() method 
class_label_predictions = model.predict(X_test_feature_vector)

# 4. Compute the Area Under the ROC curve (AUC) for the test data. Note that this time we are using one 
# function 'roc_auc_score()' to compute the auc rather than using both 'roc_curve()' and 'auc()' as we have 
# done in the past
auc = roc_auc_score(y_test, probability_predictions)
print('AUC on the test data: {:.4f}'.format(auc))

AUC on the test data: 0.9876


Let's check two emails and see if our model properly predicted whether an email is spam or not spam.

In [31]:
print('Email #1:\n')
print(original_X[14])

print('\nPrediction: Is this a spam email? {}\n'.format(class_label_predictions[14])) 

print('Actual: Is this a spam email? {}\n'.format(y_test.to_numpy()[14]))


Email #1:

Subject: tenaska iv july
darren :
please remove the price on the tenaska iv sale , deal 384258 , for july and enter the demand fee . the amount should be $ 3 , 902 , 687 . 50 .
thanks ,
megan

Prediction: Is this a spam email? True

Actual: Is this a spam email? True



In [32]:
print('Email #2:\n')
print(original_X[132])

print('\nPrediction: Is this a spam email? {}\n'.format(class_label_predictions[132])) 

print('Actual: Is this a good spam email? {}\n'.format(y_test.to_numpy()[132]))

Email #2:

Subject: re : noms / actual flow for 3 / 29 / 01
we agree with the nomination .
" eileen ponton " on 03 / 30 / 2001 10 : 05 : 40 am
to : david avila / lsp / enserch / us @ tu , charlie stone / texas utilities @ tu , melissa
jones / texas utilities @ tu , hpl . scheduling @ enron . com ,
liz . bellamy @ enron . com
cc :
subject : noms / actual flow for 3 / 29 / 01
nom mcf mmbtu
24 , 583 24 , 999 25 , 674
btu = 1 . 027

Prediction: Is this a spam email? False

Actual: Is this a good spam email? False

